[Performane] : Fix slow loading of Music Blocks on all the browsers

apsinghdev opened the pull request:
fixes: #4305 

This PR fixes the slow loading of Music Blocks. 

(Draft)

This PR will contain comparatively bigger changes. So I'll keep on adding commits here to describe what they do.

### Current Stats

<img width="741" alt="Screenshot 2025-02-01 at 2 02 51 PM" src="https://github.com/user-attachments/assets/f08ab981-3320-4824-94cb-13340810ebd9" />

Total loading time - `~ 16.0 seconds`

cc @walterbender @pikurasa 

### Changes

82bf4ca7460365be83c96b288d20274e3b9a5088 formats the index.html file as it was not formatted properly.
4ffa1342628812d82c1cb13c02a054b3fc362533 add header links in optimized order to efficient loading
10723b39c305f9df758c4854f30ba7cf1e555e17 fixes the `render-blocking` issue and bring the `cumulative-layout-shift` score from `0.20` to `0.07`
5084cf7dd7e35209d21744e912e39d62015f64f1 implements `http-server` replacing the `SimpleHTTPServer`. Upgrades `HTTP` from `1.0` to `1.1` with `Keep-Alive` enabled
e6ad80f4a9a0339b68effe114f61a606dab2dbd8 reduces `Largest Contentful Paint element` from `4,297 ms` to `2,400 ms`

omsuneri commented:
@apsinghdev I would love to contribute to this PR can you suggest me some reference

apsinghdev commented:
> @apsinghdev I would love to contribute to this PR can you suggest me some reference

Thanks @omsuneri. I'll let you know if I need help. 🙂

walterbender commented:
This is already a big improvement.

One minor issue: the splash screen is no longer centered horizontally.

![Screenshot From 2025-02-01 11-29-38](https://github.com/user-attachments/assets/eac9dcd1-2bf1-4cb3-935e-560bcc4db1dc)

apsinghdev commented:
@walterbender Thanks for letting me know. Don't know why it didn't appear on my browser. Anyways, next task is to optimise this video as it's blocking the other ops. It's causing `Largest-content-page` to load in around `4530 ms`.

apsinghdev commented:
Update:

fixed the issue: https://github.com/sugarlabs/musicblocks/pull/4331#issuecomment-2629017616

Experimented running MB using `http-server`

We are using SimpleHTTPServer to run Music Blocks to the client. But, SimpleHTTPServer doesn't support advanced features like Keep-Alive. I have experimented with running MB using http-server and it has Keep-Alive by default. 

<img width="799" alt="Screenshot 2025-02-15 at 9 22 14 PM" src="https://github.com/user-attachments/assets/5d719ede-6e86-47d9-84e6-785c894fe9e3" />


Also, I found http-server is comparatively fast increasing performance scores from 70 to 75 on the lighthouse.

<img width="624" alt="Screenshot 2025-02-15 at 9 30 06 PM" src="https://github.com/user-attachments/assets/4d05d6b4-8199-47d6-93e7-1d5a3ac5fdcb" />

Getting some feedback on these ideas. Will push it soon!

cc @walterbender @pikurasa

walterbender commented:
maybe https-server?

apsinghdev commented:
> maybe https-server?

I've replaced `SimpleHTTPServer` with `http-server`. Also as we don't need python now to run MB, I have made some changes in docs as well. Please test this commit https://github.com/sugarlabs/musicblocks/pull/4331/commits/5084cf7dd7e35209d21744e912e39d62015f64f1 to your side.

apsinghdev commented:
Interesting idea!

@walterbender Why not use a custom server using `express` that gives us the freedom to customize things as we need and make things really fast with its default features?

I tested it and it is showing really good performance.

<img width="617" alt="Screenshot 2025-02-16 at 2 28 27 PM" src="https://github.com/user-attachments/assets/3d875508-1968-4085-963c-c00ba90e55c8" />

If you get a chance, please test https://github.com/sugarlabs/musicblocks/pull/4331/commits/62397de7f799ee0a186573a9102e0ce2c66241ee

walterbender commented:
Seems good to me.

apsinghdev commented:
> Seems good to me.

Thanks! I'll do more testing.

apsinghdev commented:
@walterbender, I've fixed most of the critical issues blocking MB's loading process. I've noticed two main issues that need to be worked on:

1. Unoptimized DOM-related code: we've lots of unoptimized code that interacts with DOM leading to delays in loading.
2. Outdated/Unoptimized libraries: Some of the libraries we are using contain unused/deprecated code and it slows the performance.

To tackle both of these issues we'd need to plan about it as these are big breaking changes that will impact the whole codebase.

For now, I've made this PR read-for-review and will work on it as we discuss. Please have a look.

pikurasa commented:
@apsinghdev The video perfectly fills the screen now. However, the loading text is now missing.

As for load time, I tested the master branch and https://github.com/sugarlabs/musicblocks/commit/62397de7f799ee0a186573a9102e0ce2c66241ee with the default Sol, Mi, Sol code. Both took 13 seconds to load.

apsinghdev commented:
@pikurasa Thanks for your feedback! I'll fix the missing loading text issue. Also, which tool you used to calculate the load time?

pikurasa commented:
> @pikurasa Thanks for your feedback! I'll fix the missing loading text issue. Also, which tool have you used to calculate the load time?

A simple phone timer.

apsinghdev commented:
@pikurasa I have tested the changes and it's varying from `9 to 12` seconds. Caching is also a factor that affects the speed on different machines. Also the major reason that is slowing down the loading is the one that I've mentioned in this comment https://github.com/sugarlabs/musicblocks/pull/4331#issuecomment-2664937997 -Unoptimized DOM. Probably, I'll work on it in another PR.

apsinghdev commented:
@pikurasa I have fixed the loading text issue.
![Uploading Screenshot 2025-02-25 at 10.52.14 PM.png…]()

apsinghdev commented:
@walterbender I guess, this PR is ready to be merged. Will create a new ticket for DOM improvements and will raise PRs in that. Please let me know if I need to change something.

apsinghdev commented:
@walterbender Is there any feedback on this PR?

walterbender commented:
This seems to really help with loading times. I need to do some more testing, but all-in-all, big improvement. Thx.

pikurasa commented:
I tested it, and it works.

As for performance, it seems to improve load time by a second, at least for running it locally. (I wish I had a more robust way to test this for you...)

apsinghdev commented:
@walterbender @pikurasa  Thank you! I'll perform more tests while working on the DOM issue as both of these issues are closely related. Also, this PR has addressed the issues that were needed to be resolved before working on the DOM issue.
